UnitTests/GitUITests/UITest.cs

```diff
@@ -27,29 +29,41 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
+            Action showForm,
             Func<T, Task> runAsync)
             where T : Form
         {
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
             // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
             // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
             // complete.
             var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
+                // Wait for the form to be opened by the test thread.
+                T form;
+                do
+                {
+                    await WaitForIdleAsync(); // does also SwitchToMainThreadAsync
+                    form = Application.OpenForms.OfType<T>().SingleOrDefault();
+                }
+                while (form == null || !form.Visible);
```

Comment:
❓ In what specific case can this fail?

📝 This `do`/`while` is not required by contract of `WaitForIdleAsync`. There should not be a change to this block of code if `WaitForIdleAsync` is behaving correctly, so I would need specifics regarding the sequence that could lead to this failure.

Reply:
It is required. This race condition was causing the hangs:
There is no guarantee that `WaitForIdleAsync` is called late enough.
`showForm` must already have begun to open the form. But if it does preparation steps, `WaitForIdleAsync` could have been called while the message queue is still empty.

Reply:
works without the `do { } while` loop now